src/psyclone/psyir/nodes/node.py

```diff
@@ -1629,6 +1629,31 @@ def debug_string(self):
         from psyclone.psyir.backend.debug_writer import DebugWriter
         return DebugWriter()(self)
 
+    def origin_string(self):
+        ''' Generates a string with the available information about where
+        this node has been created. It currently only works with Fortran
+        Statements or subchildren of them.
+
+        :returns: a string specifing the origin of this node.
+        :rtype: str
+        '''
+        name = self.coloured_name(False)
+        line_span = "<unknown>"
+        original_src = "<unknown>"
+        filename = "<unknown>"
+        from psyclone.psyir.nodes.statement import Statement
+        node = self if isinstance(self, Statement) else self.ancestor(Statement)
+        # Try to populate the line/src/filename
+        if node and node._ast:
```

Comment:
Although this is OK as a first cut, it brings in intimate details of the fparser parse tree which would probably best be isolated in the `fparser2` frontend.  That way, any frontend could provide a `get_origin_string()` or similar. We could query the type of `._ast` to determine which frontend we should use?

Reply:
@arporter I agree with the problem, but I am unsure about the solution.

Currently I am using hasattr and the knowledge of fparser tree composition, which is not ideal.

> We could query the type of ._ast to determine which frontend we should use?

However, if I do that the logic will be in fparser2.py but I will have to import fparser2 library here to compare the type of the _ast node, creating a real dependency to fparser2, which I want to avoid.

I also want to avoid creating the origin string pre-emptively and store it in the node because that is a lot of string manipulation for something that only is needed very occasionally.

Maybe the solution is to add an optional _generate_origin_string attribute to the PSyIR node which is initialised with a Callable (a closure) that generates the string and will allow us to remove the _ast itself. This is a significant change since it requires to update all the fparser2.py PSyIR node creations and it starts complicating things a bit, so I wanted to check with you what do you think

Reply:
> However, if I do that the logic will be in fparser2.py but I will have to import fparser2 library here to compare the type of the _ast node, creating a real dependency to fparser2, which I want to avoid.

Could we tag the generated PSyIR with the identity of the frontend used to create it?

Reply:
Implementing this would need a bit of experimentation, but Grenoble already uses this feature and I would like them to follow master and not a specific branch, so for the time being a renamed this PR to a towards and added a TODO with where we are in this discussion. Let me know if thats ok.

Reply:
Yes, I think it's OK for now - it's not as though we have a lot of frontends :-)